File: parsers/typescript/src/parser.ts
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 MuVeraAI Corporation

import type {
  AgentIdentification,
  AgentsPolicy,
  DataHandling,
  IdentitySection,
  ParseError,
  ParseResult,
  ParseWarning,
  RateLimits,
  Restrictions,
  TrustRequirements,
} from './types.js';

/**
 * Internal representation of a parsed section from the AGENTS.md file.
 * Maps lowercase keys to their raw string values.
 */
type SectionData = Map<string, string>;

/**
 * Parses AGENTS.md content into a structured AgentsPolicy.
 *
 * Usage:
 * ```typescript
 * const parser = new AgentsMdParser();
 * const result = parser.parse(fileContent);
 * if (result.success && result.policy) {
 *   console.log(result.policy.identity.site);
 * }
 * ```
 */
export class AgentsMdParser {
  /**
   * Parses the raw string content of an AGENTS.md file.
   *
   * @param content - The raw UTF-8 string content of the AGENTS.md file.
   * @returns A ParseResult containing the policy (on success), errors, and warnings.
   */
  public parse(content: string): ParseResult {
    const errors: ParseError[] = [];
    const warnings: ParseWarning[] = [];

    if (!content || content.trim().length === 0) {
      return {
        success: false,
        errors: [{ section: 'file', message: 'File is empty.' }],
        warnings: [],
      };
    }

    const sections = this.extractSections(content);

    // Identity is the only REQUIRED section.
    if (!sections.has('identity')) {
      return {
        success: false,
        errors: [
          {
            section: 'identity',
            message:
              'Missing required ## Identity section. An AGENTS.md file must contain an Identity section.',
          },
        ],
        warnings: [],
      };
    }

    const identityData = sections.get('identity')!;
    const identityResult = this.parseIdentitySection(identityData, warnings);
    if (!identityResult) {
      return {
        success: false,
        errors: [
          {
            section: 'identity',
            message: 'The Identity section is missing the required "site" key.',
          },
        ],
        warnings,
      };
    }

    const trustRequirements = this.parseTrustRequirements(
      sections.get('trust requirements'),
      warnings,
    );
    const allowedActions = this.parseAllowedActions(
      sections.get('allowed actions'),
      warnings,
    );
    const rateLimits = this.parseRateLimits(sections.get('rate limits'), warnings);
    const dataHandling = this.parseDataHandling(sections.get('data handling'), warnings);
    const restrictions = this.parseRestrictions(sections.get('restrictions'), warnings);
    const agentIdentification = this.parseAgentIdentification(
      sections.get('agent identification'),
      warnings,
    );

    const policy: AgentsPolicy = {
      identity: identityResult,
      trustRequirements,
      allowedActions,
      rateLimits,
      dataHandling,
      restrictions,
      agentIdentification,
    };

    return {
      success: errors.length === 0,
      policy,
      errors,
      warnings,
    };
  }

  /**
   * Splits the file content into named sections by level-2 Markdown headings.
   * Section names are normalized to lowercase.
   *
   * @param content - Raw file content.
   * @returns A Map from normalized section name to the section's raw text content.
   */
  private extractSections(content: string): Map<string, SectionData> {
    const result = new Map<string, SectionData>();
    const lines = content.split(/\r?\n/);
    let currentSectionName: string | null = null;
    let currentSectionLines: string[] = [];

    const flushSection = (): void => {
      if (currentSectionName !== null) {
        result.set(currentSectionName, this.parseKeyValueLines(currentSectionLines));
        currentSectionLines = [];
      }
    };

    for (const line of lines) {
      const sectionMatch = line.match(/^##\s+(.+)$/);
      if (sectionMatch) {
        flushSection();
        currentSectionName = sectionMatch[1].trim().toLowerCase();
        currentSectionLines = [];
        continue;
      }

      if (currentSectionName !== null) {
        currentSectionLines.push(line);
      }
    }

    flushSection();
    return result;
  }

  /**
   * Parses the lines within a section body into a key-value map.
   * Only lines beginning with "- " are treated as directives.
   * Keys are normalized to lowercase and trimmed.
   * Values are trimmed.
   *
   * @param lines - The raw lines of a section body.
   * @returns A Map from normalized key to raw value string.
   */
  private parseKeyValueLines(lines: string[]): SectionData {
    const data: SectionData = new Map();
    for (const line of lines) {
      const trimmed = line.trim();
      // Skip blank lines and comment lines.
      if (!trimmed || trimmed.startsWith('#')) continue;
      // Only process directive lines starting with "- ".
      if (!trimmed.startsWith('- ')) continue;

      const withoutBullet = trimmed.slice(2);
      const colonIndex = withoutBullet.indexOf(':');
      if (colonIndex === -1) continue;

      const key = withoutBullet.slice(0, colonIndex).trim().toLowerCase();
      const value = withoutBullet.slice(colonIndex + 1).trim();

      if (key.length > 0) {
        data.set(key, value);
      }
    }
    return data;
  }

  /**
   * Converts a raw string value to a boolean.
   * Returns undefined and pushes a warning if the value is not recognized.
   */
  private parseBooleanValue(
    raw: string,
    section: string,
    key: string,
    warnings: ParseWarning[],
  ): boolean | undefined {
    const normalized = raw.toLowerCase().trim();
    if (['true', 'yes', '1', 'on'].includes(normalized)) return true;
    if (['false', 'no', '0', 'off'].includes(normalized)) return false;

    warnings.push({
      section,
      message: `Unrecognized boolean value "${raw}" for key "${key}". Expected: true/false/yes/no/1/0/on/off.`,
    });
    return undefined;
  }

  /**
   * Converts a raw string value to a positive integer.
   * Returns undefined and pushes a warning if the value is not a valid integer.
   */
  private parseIntegerValue(
    raw: string,
    section: string,
    key: string,
    warnings: ParseWarning[],
  ): number | undefined {
    const parsed = parseInt(raw, 10);
    if (isNaN(parsed) || parsed.toString() !== raw.trim()) {
      warnings.push({
        section,
        message: `Invalid integer value "${raw}" for key "${key}".`,
      });
      return undefined;
    }
    return parsed;
  }

  /**
   * Converts a comma-separated string into an array of trimmed, non-empty strings.
   */
  private parseArrayValue(raw: string): string[] {
    return raw
      .split(',')
      .map((item) => item.trim())
      .filter((item) => item.length > 0);
  }

  /**
   * Parses the Identity section.
   * Returns null if the required "site" key is absent.
   */
  private parseIdentitySection(
    data: SectionData,
    warnings: ParseWarning[],
  ): IdentitySection | null {
    const site = data.get('site');
    if (!site || site.trim().length === 0) {
      return null;
    }

    const identity: IdentitySection = { site: site.trim() };

    const contact = data.get('contact');
    if (contact) identity.contact = contact;

    const lastUpdated = data.get('last-updated');
    if (lastUpdated) identity.lastUpdated = lastUpdated;

    const specVersion = data.get('spec-version');
    if (specVersion) identity.specVersion = specVersion;

    // Warn about unrecognized keys.
    const knownKeys = new Set(['site', 'contact', 'last-updated', 'spec-version']);
    for (const key of data.keys()) {
      if (!knownKeys.has(key) && !key.startsWith('x-')) {
        warnings.push({
          section: 'identity',
          message: `Unrecognized key "${key}" in Identity section.`,
        });
      }
    }

    return identity;
  }

  /**
   * Parses the Trust Requirements section.
   * Returns specification defaults if the section is absent.
   */
  private parseTrustRequirements(
    data: SectionData | undefined,
    warnings: ParseWarning[],
  ): TrustRequirements {
    const defaults: TrustRequirements = {
      minimumTrustLevel: 0,
      authentication: 'none',
    };

    if (!data) return defaults;

    const result: TrustRequirements = { ...defaults };

    const rawLevel = data.get('minimum-trust-level');
    if (rawLevel !== undefined) {
      const level = this.parseIntegerValue(
        rawLevel,
        'trust requirements',
        'minimum-trust-level',
        warnings,
      );
      if (level !== undefined) {
        if (level < 0 || level > 5) {
          warnings.push({
            section: 'trust requirements',
            message: `Trust level ${level} is outside the valid 0-5 range. Clamping to nearest valid value.`,
          });
          result.minimumTrustLevel = Math.max(0, Math.min(5, level));
        } else {
          result.minimumTrustLevel = level;
        }
      }
    }

    const rawAuth = data.get('authentication');
    if (rawAuth !== undefined) {
      const normalized = rawAuth.toLowerCase().trim();
      if (normalized === 'required' || normalized === 'optional' || normalized === 'none') {
        result.authentication = normalized;
      } else {
        warnings.push({
          section: 'trust requirements',
          message: `Unrecognized authentication value "${rawAuth}". Expected: required/optional/none.`,
        });
      }
    }

    const rawMethods = data.get('authentication-methods');
    if (rawMethods !== undefined) {
      result.authenticationMethods = this.parseArrayValue(rawMethods);
    }

    return result;
  }

  /**
   * Parses the Allowed Actions section.
   * Returns defaults (readContent: true, all others: false) when the section is absent.
   */
  private parseAllowedActions(
    data: SectionData | undefined,
    warnings: ParseWarning[],
  ): Record<string, boolean> {
    const defaults: Record<string, boolean> = {
      readContent: true,
      submitForms: false,
      makePurchases: false,
      modifyAccount: false,
      accessApi: false,
      downloadFiles: false,
      uploadFiles: false,
      sendMessages: false,
      deleteData: false,
      createContent: false,
    };

    if (!data) return defaults;

    // Map from markdown key format to camelCase property name.
    const keyMap: Record<string, string> = {
      'read-content': 'readContent',
      'submit-forms': 'submitForms',
      'make-purchases': 'makePurchases',
      'modify-account': 'modifyAccount',
      'access-api': 'accessApi',
      'download-files': 'downloadFiles',
      'upload-files': 'uploadFiles',
      'send-messages': 'sendMessages',
      'delete-data': 'deleteData',
      'create-content': 'createContent',
    };

    const result: Record<string, boolean> = { ...defaults };

    for (const [rawKey, rawValue] of data.entries()) {
      const camelKey = keyMap[rawKey] ?? this.kebabToCamel(rawKey);
      const boolValue = this.parseBooleanValue(
        rawValue,
        'allowed actions',
        rawKey,
        warnings,
      );
      if (boolValue !== undefined) {
        result[camelKey] = boolValue;
      }
    }

    return result;
  }

  /**
   * Converts a kebab-case string to camelCase.
   * Used for custom action keys not in the standard key map.
   */
  private kebabToCamel(kebab: string): string {
    return kebab.replace(/-([a-z])/g, (_, char: string) => char.toUpperCase());
  }

  /**
   * Parses the Rate Limits section.
   * Returns an empty object (no limits) if the section is absent.
   */
  private parseRateLimits(
    data: SectionData | undefined,
    warnings: ParseWarning[],
  ): RateLimits {
    if (!data) return {};

    const result: RateLimits = {};

    const rawRpm = data.get('requests-per-minute');
    if (rawRpm !== undefined) {
      const value = this.parseIntegerValue(
        rawRpm,
        'rate limits',
        'requests-per-minute',
        warnings,
      );
      if (value !== undefined) result.requestsPerMinute = value;
    }

    const rawRph = data.get('requests-per-hour');
    if (rawRph !== undefined) {
      const value = this.parseIntegerValue(
        rawRph,
        'rate limits',
        'requests-per-hour',
        warnings,
      );
      if (value !== undefined) result.requestsPerHour = value;
    }

    const rawConcurrent = data.get('concurrent-sessions');
    if (rawConcurrent !== undefined) {
      const value = this.parseIntegerValue(
        rawConcurrent,
        'rate limits',
        'concurrent-sessions',
        warnings,
      );
      if (value !== undefined) result.concurrentSessions = value;
    }

    return result;
  }

  /**
   * Parses the Data Handling section.
   * Returns an empty partial object if the section is absent.
   */
  private parseDataHandling(
    data: SectionData | undefined,
    warnings: ParseWarning[],
  ): Partial<DataHandling> {
    if (!data) return {};

    const result: Partial<DataHandling> = {};

    const validCollectionValues = ['none', 'minimal', 'standard', 'extensive'] as const;
    const rawCollection = data.get('personal-data-collection');
    if (rawCollection !== undefined) {
      const normalized = rawCollection.toLowerCase().trim() as (typeof validCollectionValues)[number];
      if (validCollectionValues.includes(normalized)) {
        result.personalDataCollection = normalized;
      } else {
        warnings.push({
          section: 'data handling',
          message: `Unrecognized personal-data-collection value "${rawCollection}". Expected: none/minimal/standard/extensive.`,
        });
      }
    }

    const validRetentionValues = ['none', 'session-only', '30-days', '1-year', 'indefinite'] as const;
    const rawRetention = data.get('data-retention');
    if (rawRetention !== undefined) {
      const normalized = rawRetention.toLowerCase().trim() as (typeof validRetentionValues)[number];
      if (validRetentionValues.includes(normalized)) {
        result.dataRetention = normalized;
      } else {
        warnings.push({
          section: 'data handling',
          message: `Unrecognized data-retention value "${rawRetention}". Expected: none/session-only/30-days/1-year/indefinite.`,
        });
      }
    }

    const validSharingValues = ['none', 'anonymized', 'with-consent', 'unrestricted'] as const;
    const rawSharing = data.get('third-party-sharing');
    if (rawSharing !== undefined) {
      const normalized = rawSharing.toLowerCase().trim() as (typeof validSharingValues)[number];
      if (validSharingValues.includes(normalized)) {
        result.thirdPartySharing = normalized;
      } else {
        warnings.push({
          section: 'data handling',
          message: `Unrecognized third-party-sharing value "${rawSharing}". Expected: none/anonymized/with-consent/unrestricted.`,
        });
      }
    }

    const rawGdpr = data.get('gdpr-compliance');
    if (rawGdpr !== undefined) {
      const value = this.parseBooleanValue(
        rawGdpr,
        'data handling',
        'gdpr-compliance',
        warnings,
      );
      if (value !== undefined) result.gdprCompliance = value;
    }

    return result;
  }

  /**
   * Parses the Restrictions section.
   * Returns empty arrays if the section is absent.
   */
  private parseRestrictions(
    data: SectionData | undefined,
    warnings: ParseWarning[],
  ): Restrictions {
    if (!data) {
      return {
        disallowedPaths: [],
        requireHumanApproval: [],
        readOnlyPaths: [],
      };
    }

    const parsePaths = (raw: string | undefined): string[] => {
      if (!raw) return [];
      return this.parseArrayValue(raw);
    };

    const restrictions: Restrictions = {
      disallowedPaths: parsePaths(data.get('disallowed-paths')),
      requireHumanApproval: parsePaths(data.get('require-human-approval')),
      readOnlyPaths: parsePaths(data.get('read-only-paths')),
    };

    // Warn about paths that don't start with /.
    const allPaths = [
      ...restrictions.disallowedPaths,
      ...restrictions.requireHumanApproval,
      ...restrictions.readOnlyPaths,
    ];
    for (const path of allPaths) {
      if (!path.startsWith('/')) {
        warnings.push({
          section: 'restrictions',
          message: `Path pattern "${path}" does not start with "/". Path patterns should be absolute paths.`,
        });
      }
    }

    return restrictions;
  }

  /**
   * Parses the Agent Identification section.
   * Returns permissive defaults if the section is absent.
   */
  private parseAgentIdentification(
    data: SectionData | undefined,
    warnings: ParseWarning[],
  ): AgentIdentification {
    const defaults: AgentIdentification = {
      requireAgentHeader: false,
      requireDisclosure: false,
    };

    if (!data) return defaults;

    const result: AgentIdentification = { ...defaults };

    const rawRequireHeader = data.get('require-agent-header');
    if (rawRequireHeader !== undefined) {
      const value = this.parseBooleanValue(
        rawRequireHeader,
        'agent identification',
        'require-agent-header',
        warnings,
      );
      if (value !== undefined) result.requireAgentHeader = value;
    }

    const headerName = data.get('agent-header-name');
    if (headerName) result.agentHeaderName = headerName;

    const rawRequireDisclosure = data.get('require-disclosure');
    if (rawRequireDisclosure !== undefined) {
      const value = this.parseBooleanValue(
        rawRequireDisclosure,
        'agent identification',
        'require-disclosure',
        warnings,
      );
      if (value !== undefined) result.requireDisclosure = value;
    }

    return result;
  }
}

File: parsers/python/src/agents_md/parser.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2026 MuVeraAI Corporation

"""
AGENTS.md parser — converts raw AGENTS.md file content into an AgentsPolicy.

Usage:
    from agents_md import AgentsMdParser

    parser = AgentsMdParser()
    result = parser.parse(file_content)
    if result.success and result.policy:
        print(result.policy.identity.site)
"""

from __future__ import annotations

import re
from typing import Literal

from .types import (
    AgentIdentification,
    AgentsPolicy,
    DataHandling,
    IdentitySection,
    ParseError,
    ParseResult,
    ParseWarning,
    RateLimits,
    Restrictions,
    TrustRequirements,
)

# Type aliases used internally.
SectionData = dict[str, str]

# Keys for the Allowed Actions section (markdown key -> Python attribute name).
_ACTION_KEY_MAP: dict[str, str] = {
    "read-content": "read_content",
    "submit-forms": "submit_forms",
    "make-purchases": "make_purchases",
    "modify-account": "modify_account",
    "access-api": "access_api",
    "download-files": "download_files",
    "upload-files": "upload_files",
    "send-messages": "send_messages",
    "delete-data": "delete_data",
    "create-content": "create_content",
}

_DEFAULT_ALLOWED_ACTIONS: dict[str, bool] = {
    "read_content": True,
    "submit_forms": False,
    "make_purchases": False,
    "modify_account": False,
    "access_api": False,
    "download_files": False,
    "upload_files": False,
    "send_messages": False,
    "delete_data": False,
    "create_content": False,
}


class AgentsMdParser:
    """
    Parses AGENTS.md content into a structured AgentsPolicy.

    This parser is stateless — each call to ``parse()`` is independent.
    """

    def parse(self, content: str) -> ParseResult:
        """
        Parse the raw string content of an AGENTS.md file.

        Args:
            content: The raw UTF-8 string content of the AGENTS.md file.

        Returns:
            A ParseResult containing the policy (on success), errors, and warnings.
        """
        errors: list[ParseError] = []
        warnings: list[ParseWarning] = []

        if not content or not content.strip():
            return ParseResult(
                success=False,
                errors=[ParseError(section="file", message="File is empty.")],
            )

        sections = self._extract_sections(content)

        if "identity" not in sections:
            return ParseResult(
                success=False,
                errors=[
                    ParseError(
                        section="identity",
                        message=(
                            "Missing required ## Identity section. "
                            "An AGENTS.md file must contain an Identity section."
                        ),
                    )
                ],
            )

        identity = self._parse_identity_section(sections["identity"], warnings)
        if identity is None:
            return ParseResult(
                success=False,
                errors=[
                    ParseError(
                        section="identity",
                        message='The Identity section is missing the required "site" key.',
                    )
                ],
                warnings=warnings,
            )

        trust_requirements = self._parse_trust_requirements(
            sections.get("trust requirements"), warnings
        )
        allowed_actions = self._parse_allowed_actions(
            sections.get("allowed actions"), warnings
        )
        rate_limits = self._parse_rate_limits(sections.get("rate limits"), warnings)
        data_handling = self._parse_data_handling(sections.get("data handling"), warnings)
        restrictions = self._parse_restrictions(sections.get("restrictions"), warnings)
        agent_identification = self._parse_agent_identification(
            sections.get("agent identification"), warnings
        )

        policy = AgentsPolicy(
            identity=identity,
            trust_requirements=trust_requirements,
            allowed_actions=allowed_actions,
            rate_limits=rate_limits,
            data_handling=data_handling,
            restrictions=restrictions,
            agent_identification=agent_identification,
        )

        return ParseResult(
            success=len(errors) == 0,
            policy=policy,
            errors=errors,
            warnings=warnings,
        )

    # ------------------------------------------------------------------
    # Section extraction
    # ------------------------------------------------------------------

    def _extract_sections(self, content: str) -> dict[str, SectionData]:
        """
        Split the file content into named sections by level-2 Markdown headings.
        Section names are normalized to lowercase.
        """
        result: dict[str, SectionData] = {}
        current_section_name: str | None = None
        current_lines: list[str] = []

        def flush() -> None:
            if current_section_name is not None:
                result[current_section_name] = self._parse_key_value_lines(current_lines)

        for line in re.split(r"\r?\n", content):
            section_match = re.match(r"^##\s+(.+)$", line)
            if section_match:
                flush()
                current_section_name = section_match.group(1).strip().lower()
                current_lines = []
                continue

            if current_section_name is not None:
                current_lines.append(line)

        flush()
        return result

    def _parse_key_value_lines(self, lines: list[str]) -> SectionData:
        """
        Parse the lines within a section body into a key-value dict.
        Only lines beginning with "- " are treated as directives.
        Keys are normalized to lowercase and stripped.
        Values are stripped.
        """
        data: SectionData = {}
        for line in lines:
            trimmed = line.strip()
            if not trimmed or trimmed.startswith("#"):
                continue
            if not trimmed.startswith("- "):
                continue

            without_bullet = trimmed[2:]
            colon_index = without_bullet.find(":")
            if colon_index == -1:
                continue

            key = without_bullet[:colon_index].strip().lower()
            value = without_bullet[colon_index + 1 :].strip()

            if key:
                data[key] = value

        return data

    # ------------------------------------------------------------------
    # Value coercion helpers
    # ------------------------------------------------------------------

    def _parse_bool(
        self,
        raw: str,
        section: str,
        key: str,
        warnings: list[ParseWarning],
    ) -> bool | None:
        """Convert a raw string to a boolean. Returns None and appends a warning on failure."""
        normalized = raw.lower().strip()
        if normalized in ("true", "yes", "1", "on"):
            return True
        if normalized in ("false", "no", "0", "off"):
            return False
        warnings.append(
            ParseWarning(
                section=section,
                message=(
                    f'Unrecognized boolean value "{raw}" for key "{key}". '
                    "Expected: true/false/yes/no/1/0/on/off."
                ),
            )
        )
        return None

    def _parse_int(
        self,
        raw: str,
        section: str,
        key: str,
        warnings: list[ParseWarning],
    ) -> int | None:
        """Convert a raw string to an integer. Returns None and appends a warning on failure."""
        try:
            return int(raw.strip())
        except ValueError:
            warnings.append(
                ParseWarning(
                    section=section,
                    message=f'Invalid integer value "{raw}" for key "{key}".',
                )
            )
            return None

    @staticmethod
    def _parse_array(raw: str) -> list[str]:
        """Convert a comma-separated string to a list of trimmed non-empty strings."""
        return [item.strip() for item in raw.split(",") if item.strip()]

    @staticmethod
    def _kebab_to_snake(kebab: str) -> str:
        """Convert a kebab-case string to snake_case."""
        return kebab.replace("-", "_")

    # ------------------------------------------------------------------
    # Section parsers
    # ------------------------------------------------------------------

    def _parse_identity_section(
        self,
        data: SectionData,
        warnings: list[ParseWarning],
    ) -> IdentitySection | None:
        """Parse the Identity section. Returns None if the required 'site' key is absent."""
        site = data.get("site", "").strip()
        if not site:
            return None

        identity = IdentitySection(site=site)

        if contact := data.get("contact"):
            identity.contact = contact
        if last_updated := data.get("last-updated"):
            identity.last_updated = last_updated
        if spec_version := data.get("spec-version"):
            identity.spec_version = spec_version

        known_keys = {"site", "contact", "last-updated", "spec-version"}
        for key in data:
            if key not in known_keys and not key.startswith("x-"):
                warnings.append(
                    ParseWarning(
                        section="identity",
                        message=f'Unrecognized key "{key}" in Identity section.',
                    )
                )

        return identity

    def _parse_trust_requirements(
        self,
        data: SectionData | None,
        warnings: list[ParseWarning],
    ) -> TrustRequirements:
        """Parse the Trust Requirements section. Returns defaults when section is absent."""
        if data is None:
            return TrustRequirements()

        result = TrustRequirements()

        if (raw_level := data.get("minimum-trust-level")) is not None:
            level = self._parse_int(raw_level, "trust requirements", "minimum-trust-level", warnings)
            if level is not None:
                if level < 0 or level > 5:
                    warnings.append(
                        ParseWarning(
                            section="trust requirements",
                            message=(
                                f"Trust level {level} is outside the valid 0-5 range. "
                                "Clamping to nearest valid value."
                            ),
                        )
                    )
                    result.minimum_trust_level = max(0, min(5, level))
                else:
                    result.minimum_trust_level = level

        if (raw_auth := data.get("authentication")) is not None:
            normalized = raw_auth.lower().strip()
            if normalized in ("required", "optional", "none"):
                result.authentication = normalized  # type: ignore[assignment]
            else:
                warnings.append(
                    ParseWarning(
                        section="trust requirements",
                        message=(
                            f'Unrecognized authentication value "{raw_auth}". '
                            "Expected: required/optional/none."
                        ),
                    )
                )

        if (raw_methods := data.get("authentication-methods")) is not None:
            result.authentication_methods = self._parse_array(raw_methods)

        return result

    def _parse_allowed_actions(
        self,
        data: SectionData | None,
        warnings: list[ParseWarning],
    ) -> dict[str, bool]:
        """Parse the Allowed Actions section. Returns defaults when section is absent."""
        result = dict(_DEFAULT_ALLOWED_ACTIONS)

        if data is None:
            return result

        for raw_key, raw_value in data.items():
            snake_key = _ACTION_KEY_MAP.get(raw_key, self._kebab_to_snake(raw_key))
            bool_value = self._parse_bool(raw_value, "allowed actions", raw_key, warnings)
            if bool_value is not None:
                result[snake_key] = bool_value

        return result

    def _parse_rate_limits(
        self,
        data: SectionData | None,
        warnings: list[ParseWarning],
    ) -> RateLimits:
        """Parse the Rate Limits section. Returns empty RateLimits when section is absent."""
        if data is None:
            return RateLimits()

        result = RateLimits()

        if (raw := data.get("requests-per-minute")) is not None:
            result.requests_per_minute = self._parse_int(
                raw, "rate limits", "requests-per-minute", warnings
            )

        if (raw := data.get("requests-per-hour")) is not None:
            result.requests_per_hour = self._parse_int(
                raw, "rate limits", "requests-per-hour", warnings
            )

        if (raw := data.get("concurrent-sessions")) is not None:
            result.concurrent_sessions = self._parse_int(
                raw, "rate limits", "concurrent-sessions", warnings
            )

        return result

    def _parse_data_handling(
        self,
        data: SectionData | None,
        warnings: list[ParseWarning],
    ) -> DataHandling:
        """Parse the Data Handling section. Returns empty DataHandling when section is absent."""
        if data is None:
            return DataHandling()

        result = DataHandling()

        valid_collection = ("none", "minimal", "standard", "extensive")
        if (raw := data.get("personal-data-collection")) is not None:
            normalized = raw.lower().strip()
            if normalized in valid_collection:
                result.personal_data_collection = normalized  # type: ignore[assignment]
            else:
                warnings.append(
                    ParseWarning(
                        section="data handling",
                        message=(
                            f'Unrecognized personal-data-collection value "{raw}". '
                            f"Expected: {'/'.join(valid_collection)}."
                        ),
                    )
                )

        valid_retention = ("none", "session-only", "30-days", "1-year", "indefinite")
        if (raw := data.get("data-retention")) is not None:
            normalized = raw.lower().strip()
            if normalized in valid_retention:
                result.data_retention = normalized  # type: ignore[assignment]
            else:
                warnings.append(
                    ParseWarning(
                        section="data handling",
                        message=(
                            f'Unrecognized data-retention value "{raw}". '
                            f"Expected: {'/'.join(valid_retention)}."
                        ),
                    )
                )

        valid_sharing = ("none", "anonymized", "with-consent", "unrestricted")
        if (raw := data.get("third-party-sharing")) is not None:
            normalized = raw.lower().strip()
            if normalized in valid_sharing:
                result.third_party_sharing = normalized  # type: ignore[assignment]
            else:
                warnings.append(
                    ParseWarning(
                        section="data handling",
                        message=(
                            f'Unrecognized third-party-sharing value "{raw}". '
                            f"Expected: {'/'.join(valid_sharing)}."
                        ),
                    )
                )

        if (raw := data.get("gdpr-compliance")) is not None:
            value = self._parse_bool(raw, "data handling", "gdpr-compliance", warnings)
            if value is not None:
                result.gdpr_compliance = value

        return result

    def _parse_restrictions(
        self,
        data: SectionData | None,
        warnings: list[ParseWarning],
    ) -> Restrictions:
        """Parse the Restrictions section. Returns empty lists when section is absent."""
        if data is None:
            return Restrictions()

        def parse_paths(raw: str | None) -> list[str]:
            if raw is None:
                return []
            return self._parse_array(raw)

        restrictions = Restrictions(
            disallowed_paths=parse_paths(data.get("disallowed-paths")),
            require_human_approval=parse_paths(data.get("require-human-approval")),
            read_only_paths=parse_paths(data.get("read-only-paths")),
        )

        all_paths = (
            restrictions.disallowed_paths
            + restrictions.require_human_approval
            + restrictions.read_only_paths
        )
        for path in all_paths:
            if not path.startswith("/"):
                warnings.append(
                    ParseWarning(
                        section="restrictions",
                        message=(
                            f'Path pattern "{path}" does not start with "/". '
                            "Path patterns should be absolute paths."
                        ),
                    )
                )

        return restrictions

    def _parse_agent_identification(
        self,
        data: SectionData | None,
        warnings: list[ParseWarning],
    ) -> AgentIdentification:
        """Parse the Agent Identification section. Returns permissive defaults when absent."""
        if data is None:
            return AgentIdentification()

        result = AgentIdentification()

        if (raw := data.get("require-agent-header")) is not None:
            value = self._parse_bool(raw, "agent identification", "require-agent-header", warnings)
            if value is not None:
                result.require_agent_header = value

        if header_name := data.get("agent-header-name"):
            result.agent_header_name = header_name

        if (raw := data.get("require-disclosure")) is not None:
            value = self._parse_bool(raw, "agent identification", "require-disclosure", warnings)
            if value is not None:
                result.require_disclosure = value

        return result
